### ISSUE TITLE
Support Windows as a first class platform (note: not WSL or emulation)

### ISSUE THREAD

**Issue by xtagon** [crystal-lang/crystal#26]
Hi,

Crystal is an exciting project with lots of potential. I am interested in the possibility of a Crystal compiler for Windows. How feasible is that idea, and how much work would it be?

Specifically, I would like to be able to compile a Crystal program into a Windows .DLL file with exported C functions. I don't care if I run the actual Crystal compiler on a Linux or Windows machine, I just want to end up with a DLL library that will run on Windows.

I have knowledge of Ruby and some C, and I'm willing to dive into learning about LLVM if it means I can contribute here. I would be pleased if you could point me in the right direction.

Regards,
Justin

**Comment by Flaise** [comment]
Any progress on this issue? Crystal looks like exactly the kind of language I'd like to begin using for new projects but the lack of Windows support is a dealbreaker for me - Windows is where you find paying customers and I can't be investing myself in something that locks me out of having an audience.

**Comment by G4MR** [comment]
@Flaise yeah I agree, I'd love to try this on windows.

**Comment by bararchy** [comment]
So, Trying to build the Windows branch using the instructions in: https://github.com/manastech/crystal/blob/windows/README_win32.md

Step 1 produces:

```
 bin/crystal build src/compiler/crystal.cr --cross-compile "windows x86" --single-module --target "i686-pc-win32-gnu" --release -o bin/crystal
Error in ./src/file/stat.cr:81: undefined constant Stat

  fun stat(path : UInt8*, stat : Stat*) : Int32
                                 ^~~~

```

**Comment by bararchy** [comment]
Using @xwanderer fork from https://github.com/xwanderer/crystal/tree/win32
I get a similar issue regarding types

```
 bin/crystal build src/compiler/crystal.cr --cross-compile "windows x86" --single-module --target "i686-pc-win32-gnu" -o win32/crystal
Error while requiring "prelude"

in ./src/prelude.cr:17: while requiring "iterator"

require "iterator"
^

Syntax error in ./src/iterator.cr:194: type variables can only be single letters

  struct Zip(I1, I2, T1, T2)
             ^

```

**Comment by rishavs** [comment]
Just wanted to add my support to this. I really want to try out Crystal but not at the cost of my favored platform and environment.

**Comment by david50407** [comment]
And I'm going to support on this, too.
I'm trying to make the Crystal compiler run on Windows first, and go building std-lib on windows then.

**Comment by kirbyfan64** [comment]
@david50407 Well, the compiler is written in Crystal, so you kind of need to do both at the same time...maybe Cygwin could help there?

**Comment by david50407** [comment]
@kirbyfan64 I tried this way https://github.com/manastech/crystal/issues/26#issuecomment-40704994 , and run `Hello world` program well with LLVM 3.7 on Windows with TDM-GCC 5.1.0

Then I'm going to make Crystal compiler can be cross-compiled to Windows, for this, I will make a part of std-lib (core libs?) works that compiler needs.

**Comment by ozra** [comment]
So glad to see progress and interest on this front!

**Comment by mverzilli** [comment]
+1, it's great seeing someone brave enough to tackle this! <3

**Comment by anilreddy** [comment]
+1

**Comment by nob-suz** [comment]
+1 looking forward to see!, and there is recent LLVM Clang for MSVC&,NET http://www.infoworld.com/article/2994395/application-development/microsoft-taps-open-source-llvm-compiler-for-cross-platform-net.html

**Comment by david50407** [comment]
It is a problem that Windows doesn't provide some POSIX Process calls, like `fork()`, `getsid()`, ...etc.

I saw the solution in Ruby, it decide to raise `NotImplementedError` and tell user "the function is unimplemented on this machine". ( https://github.com/ruby/ruby/blob/trunk/process.c#L3690 )

Should I do the same behavior like Ruby (, throws an Exception), when I cannot find the solution on Windows? (like `Process.fork()`)
I think this will keep std-lib in the same interface (, that is, making std-lib more platform indepedent), and let coder (who codes in Crystal) to check the document and catch the Exception by their own (without causing error on compile-time).

But I don't know if raising Exception on runtime is better than throwing error on compile-time or not.
How about you guys think?

**Comment by jhass** [comment]
It should fail to compile IMO, either by calling raise in a macro or simply letting it run into an undefined call.

**Comment by lowjoel** [comment]
I think a compile-time error is more suitable, since Crystal is a compiled language, and we want to catch as many errors as early on as possible.

**Comment by david50407** [comment]
@jhass , I think raising `NotImplementedExecption` is better than undefined error.

Many languages does this on runtime, like Ruby, Python, Java, C# (even have `NotImplementedException` and `NotSupporttedException`), etc., they throw exceptions to keep the spec (or I mean the interface) of libs generally. (And it also make document auto-generating more easier.)

But we don't have VM in Crystal (, means we need to compile every time when generating other platform's executable files) , so we can just raise execption on compile-time.

And std-libs are not just C/lib bindings, they should make Crystal has a good interface to make coders build more portable applications.

**Comment by david50407** [comment]
I have one more problem when I'm trying to make compiler throws the exception on compile-time, does there any trick to show which method is 'not supported yet'?

for the code below:
```
macro raise_not_implement_yet_exception
  {% raise "Not Impl. yet" %}
end

class Foo
  def test
    raise_not_implement_yet_exception
  end
end

Foo.new.test
```

it's error like this:
```
Error in ./not_impl.cr:11: instantiating 'Foo#test()'

Foo.new.test
        ^~~~

in ./not_impl.cr:7: expanding macro

  raise_not_implement_yet_exception
  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

in ./not_impl.cr:2: can't expand macro: Not Impl. yet

 {% raise "Not Impl. yet" %}
    ^~~~~
```

I think it doesn't point out which method is 'Not Impl. yet' clearly, and it looks like it's just a macro expansion error.

**Comment by bcardiff** [comment]
@david50407 you can `puts` in compile time. It does looks a bit better. http://play.crystal-lang.org/#/r/ml8

[Regarding runtime vs compile time] If the unsupported operations do not compile then the specs won't compile as a whole. I would prefer to have a flag that shows a progress regarding how much of the stdlib is supported in windows. But, this might be more a issue of spec. Yet, as things are right now, I would vote for runtime exceptions for this case. In the long term a compilation errors sounds better.

**Comment by lowjoel** [comment]
I have not used much Crystal, but how about something radical -- put the platform-specific functionality in a library/module/include (sorry...) that needs to be explicitly imported? I gather most of these functionality that can't be implemented on Windows is due to it being Unix syscalls, and it's likely that this would be all implemented, or not implemented at all.

There are some edge cases, like OS X not supporting anonymous semaphores and stuff like that, but I think that's a little easier to deal with.

**Comment by david50407** [comment]
@lowjoel, In Ruby, you still can call `Process.fork()` on Windows, but it will raise an exception; But in golang, it even don't have `fork()` call!
The question is: how to divide out which is platform-independent method and which is platform-specific method?
As I said, std-lib should be the interface that you and I can use the method without worrying about if this method is platform-independent or have to call in another way on different platform.

BTW, And Crystal seems to only compile the code we called not whole modules right now.

**Comment by david50407** [comment]
@bcardiff @lowjoel How about making these not/un-supported methods warn (, or just showing some message for now,) on compile-time, and raise on runtime.
Because we have spec written in Crystal, if we want to test the spec, we have to pass the compilation first, and check these methods raise normally.

**Comment by kirbyfan64** [comment]
Well, Ruby isn't an error since nothing is an error until you try it...so it isn't the best example...

Honestly, it should be a compile-time error. I can't think of a single case that it would actually be *desired*.

**Comment by david50407** [comment]
@kirbyfan64 which one you think is better? raise an exception or just let the method undefined?

**Comment by rishavs** [comment]
I would recommend not leaving it undefined.

**Comment by ozra** [comment]
If using unsupported functionality when compiling for a certain platform it should definitely error at compile time! And as many already mentioned, preferably with a more specific error "not implemented for this platform", rather than just "method missing".

**Comment by ysbaddaden** [comment]
Raising in macros is unsupported as of now. For now, please raise a specific exception at runtime eg: NotImplementedError). Specs will then show what needs to be implemented.

If a feature is impossible for a specific platform (eg: fork) then skip the method definition, so code won't compile. Please open an issue if it's not satisfactory —we could have `@unsupported "message"` just like it was proposed to have `@deprecated "message"`, but that won't happen until we get near to 1.0

Also, please check the work by xwanderer on porting Crystal to windows, before restarting from scratch.

Last but not least, in order to avoid daily merge errors, it could be interesting to englob code, especially when the code is mostly about reimplementing than fixing for a platform:

```crystal
# src/file.cr
ifdef windows
  require " windows/file"
else
# untouched source code
end
```

**Comment by david50407** [comment]
@ysbaddaden I'm merging master into xwanderer's win32 branch and upgrading the exist code to 0.9.1, so these methods are in single file now.

I'm going to open an issue for `@unsupported "message"`, it is a great solution for marking  unsupported methods IMO.

And I'll try to divide out the platform-specific codes. Will Linux, Mac OS X, and other platform we may support later be in this form in the future?

**Comment by ysbaddaden** [comment]
Your call. A windows port is a long task. I proposed to make as minimal as possible changes to the current code that is likely to change often, so your work won't be hard, if not impossible, to merge eventually and during all the development process. 

The current code may eventually follow the same pattern, or not. That depends if other platforms are very different or if it's mostly a Windows vs Unix thing — libc bindings will eventually be generated automatically by [crystal_lib](https:/github.com/manastech/crystal_lib) from C headers so there is no need to extract them for each Unix platform.

**Comment by david50407** [comment]
We need SEH (Structured Exception Handling) for unwinding, and LLVM supported [new intrinsics for windows](http://llvm.org/releases/3.8.0/docs/ExceptionHandling.html#exception-handling-using-the-windows-runtime) from LLVM 3.8, but we are still using LLVM 3.5/3.6.

So we may wait for upgrading to 3.8/3.9 or I'll try other ways to use SEH. :cry:

**Comment by asterite** [comment]
Until we support Windows, it seems it's possible to run it in the Ubuntu bash inside Windows: https://www.reddit.com/r/crystal_programming/comments/4m9fjd/i_got_crystal_to_install_in_the_ubuntu_bash/

**Comment by david50407** [comment]
@asterite :tada: But it seems that cannot use Windows API and creating some cool GUI applications :disappointed:

**Comment by asterite** [comment]
@david50407 True. That's why I don't understand why everyone wants Crystal for Windows. We'd need to bind the whole Windows API. I believe C# is a better language for that, as it also comes with Visual Studio and you have a visual editor, etc.

**Comment by xtagon** [comment]
I think what people really want is Ruby on Windows, but compiled instead of interpreted, so they look to Crystal.

**Comment by bcardiff** [comment]
I would expect to see Windows API if crystal compiler is able to get windows native support. Not in the std, but it will. AFAIK WFP is not Windows API, so it will be only for win32-ish apps. But still a long way to go.

**Comment by BlaXpirit** [comment]
@asterite 
Why everyone wants Crystal for Windows:

To be able to run programs on 99.9% of desktop computers, not 11.2% as it is now. [[1]](https://en.wikipedia.org/wiki/Usage_share_of_operating_systems#Market_share_by_category)

And no, the fact that web applications don't require Windows to be served to Windows users does not console me. That is only one kind of application, and I am not interested in using Crystal for it anyway.

And using Windows API directly is not required to make GUI applications, that's for sure. That's not what people do these days.
Cross-platform library bindings for Crystal would start pouring in immediately.

**Comment by asterite** [comment]
What do users use Ruby for in Windows?

**Comment by david50407** [comment]
@asterite yes, C# is good on Windows, but we (maybe only me) just want this awesome language can run on Windows.
Why? Because it's so beautiful and amazing and I wants this also can be on Windows when I have to write something or create something :tada: 
BTW, I guess if there is any possible can move to Windows or even Crystal only bind the minimal Windows APIs, there will also be many shards or modules come from communities (yeah, Open Source power :heart:).
Porting to Windows doesn't needs any GUI binding requirements on Crystal, but I believe there will be many bindings come out. (like Windows only's GDI+ or others)

**Comment by david50407** [comment]
@asterite I don't think "What do we use Ruby on Windows" is the problem we don't want to move Crystal to Windows, how about "What do we use Node.JS/Python/Java or other languages on Windows?"
I know there can be many answers and many way to do a program, but we love Crystal. :heart:

**Comment by asterite** [comment]
Yeah... maybe I rushed a bit with my answer. I haven't used Windows in years and I don't have good memories of using it, so that's why I'm not interested in Windows that much, but it'll definitely happen in the future.

**Comment by kirbyfan64** [comment]
> I haven't used Windows in years and I don't have good memories of using it

You should try Windows 10. It's a pretty significant improvement over the crud MS has been releasing for the last several years. I mean, you can actually paste at the command prompt now!

**Comment by Flaise** [comment]
> I don't have good memories of using [Windows]

I feel similarly but try telling that to my target audience. Would running a Crystal app from the new Windows Bash prompt be enough to start a GUI toolkit or SDL window?

**Comment by ysbaddaden** [comment]
@asterite why do want Ruby or Crystal on OSX? I mean, you'd have to bind the whole cocoa libraries, and such. Or maybe you just happen to use OSX and not Linux and don't want to bother with VMs?

To build cross platform applications (with GUI) one either have to write Java or Python or C++. It would be very nice to have Crystal as an alternative, using for instance a wxwidgets binding, that would happen to be identical on all platforms. Neat.

@Flaize no. It's CLI only and it can't access windows API.

**Comment by asterite** [comment]
You all have very good points. I'm convinced now that supporting Windows would be **very** nice to have :-)

**Comment by david50407** [comment]
BTW, Visual studio now provides APIs for developers to make other language's integrations, Node.js and Python now have support by Microsoft officially [[1]](https://github.com/Microsoft/PTVS) [[2]](https://github.com/Microsoft/nodejstools) (was come out from communities).
If Crystal is going to land on Windows, it can also work with VS well. (by writing some VS extensions to interact with Crystal)
And, @asterite, we know that porting to other platforms is kind a hard work, thank to every contributors.

**Comment by ozra** [comment]
I haven't coded specifically for Win-UI in 15 years (and then only for clients, not my personal want), most of the time been using SDL, etc. Personally I loathe windows (I only have it on the machines in my music studio, clients usually have/want recordings in Cubase or so), but I think it's _very_ important that Crystal is available for it. Especially as @Flaise says: the target audience for a certain app doesn't necessarily share the values of the dev'er ;-)

There are many alternatives for UI: GTK, Qt, QT-QML and many others. Some go coders seem to go the QML-route, making the backend and frontend separate (like a web-app but QML instead), but there is a c-wrapper lib so it could probably be compiled in.

Anyways, that's that for UI's.

**Comment by rdp** [comment]
@Flaise I doubt it unless you have a working X windows server going as well (though I haven't checked into it, but might be windows 10 only, etc. etc.) but maybe work, that would be a convenient win :)

**Comment by omarroth** [comment]
I'm working on porting more complex programs to windows, and I'd like to present what I have so far. There are several windows ports of linux functionality, so there shouldn't be a copious amount of work writing platform-specific code. Binding to the windows API doesn't seem to be a problem. A sample program like this:
```
@[Link(ldflags: "-include windows.h")]
lib LibWin
  fun message_beep = MessageBeep (uType : UInt32) : Nil
end
```
appears to work fine. I've made a few changes so that everything will link correctly. Compiling a [smaller standard lib](https://gist.github.com/omarroth/64e9132ef2360e724cf7ddfde38f53a2) produces
```
$ gcc main.o -include windows.h -lgc-lib -lpthread -levent -lgw32c -L/usr/lib -L/usr/local/lib
Warning: resolving _GetLastError by linking to _GetLastError@0
Use --enable-stdcall-fixup to disable these warnings
Use --disable-stdcall-fixup to disable these fixups
main.o:(.text+0x161): undefined reference to `_tls_array'
main.o:(.text+0x196): undefined reference to `_tls_array'
main.o:(.text+0x7d1): undefined reference to `_tls_array'
main.o:(.text+0x7f4): undefined reference to `_tls_array'
main.o:(.text+0x812): undefined reference to `_tls_array'
main.o:(.text+0x84c): more undefined references to `_tls_array' follow
main.o:(.text+0x2c8d): undefined reference to `iconv_open'
main.o:(.text+0x38ec): undefined reference to `iconv_close'
main.o:(.text+0x3ad0): undefined reference to `iconv'
main.o:(.text+0x59c0): undefined reference to `GC_pthread_detach'
collect2.exe: error: ld returned 1 exit status
```
I'm having trouble finding where `tls_array` is called by the compiler, and the llvm docs don't seem to provide very much information on it either. I'm also not sure why `GC_pthread_detach` isn't being provided by gcc, although it's probably because I'm using MSYS for crystal but used VS to build bdwgc, but all the other GC functions appear to work fine, so that's something.
I'm using LLVM 3.6 for both windows and linux build environments.

**Comment by kirbyfan64** [comment]
> because I'm using MSYS for crystal but used VS to build bdwgc

Probably just the wrong bdwgc version. AFAIK all Windows compilers are C ABI compatible; it's with C++ that things get hairy.

**Comment by ysbaddaden** [comment]
Because pthreads are only available on POSIX platforms (hence the "p" I believe). You need to write an alternative Thread and GC integration specific to windows, using windows threads and whatever boehm-gc is providing.

**Comment by ysbaddaden** [comment]
Errata: the p stands for portable, but I don't think it includes anything but POSIX.

**Comment by rdp** [comment]
I typically use pthreads-win32 as a wrapper for it, though it lacks a few
things, perhaps it has enough...

**Comment by asterite** [comment]
@omarroth `tls_array` seems to be a call LLVM is inserting for generating code for thread local storage, only for Windows. I don't know why, and what's that call in windows.

**Comment by david50407** [comment]
@omarroth did you solve the problem with error handling?

**Comment by omarroth** [comment]
@david50407 I think what I have should work, but I can't really test it until I can produce a working binary.

**Comment by dimavs** [comment]
I've managed to compile simple
`puts "Hello!"`
using mingw-w64

Linux box:
 - added 'x86_64-w64-mingw' target to lib_c (pretty much x86_64-linux-gnu copy);
 - dumped llvm byte code (_DUMP=1 ../crystal/.build/crystal build test.cr --cross-compile --target "x86_64-w64-mingw" --single-module 2> test.bc_)

Windows box:
 - installed msys2/mingw-w64 environment
 - _llc test.bc_
 - _clang test.s_
 - built mmap (https://github.com/witwall/mman-win32)
 - _clang test.o stub.c -static -levent -lgc -lpthread -liconv -lws2_32 -lpcre -I. -L. -lmman_

_stub.c_ just a simple stub for _setup_sigfault_handler, fcntl, dladdr, pipe, waitpid_ (and I've actually commented out _abort_ on _fork_ check in _libgc_)

So, obviously, to support Windows all those famous _fork_ and _fcntl_ should be implemented. Not sure what status of _gnuwin32_ library is.
Also, just wondering what target we should point: mingw, cygwin or msvc?

**Comment by bcardiff** [comment]
IMO we should point for msvc. I feel is the more natural option for a windows developer.

Thanks for investing efforts for this

**Comment by omarroth** [comment]
@dimavs What happens when you try compiling something bigger? Maybe something like [samples/sieve.cr](https://github.com/crystal-lang/crystal/blob/master/samples/sieve.cr). Or the compiler itself?

@bcardiff msvc seems reasonable. Any windows developers that would like to put in their 2c?

**Comment by sumproxy** [comment]
I'd vote for mingw although any windows support is really very much appreciated!

**Comment by RX14** [comment]
The final target should be windows without any external toolchains needed (i'm assuming that's what msvc is), although mingw or cygwin would be a useful first step that could get merged.

Would adding mingw or cygwin support help with "native" windows support?

**Comment by konovod** [comment]
Cygwin is terrible. It's more like emulator then native program. Personally i'll prefer mingw, but msvc is fine too.

**Comment by omarroth** [comment]
@RX14 Well, Windows doesn't come with a C compiler like Linux does. msvc is comparable to xcode, which provides tools for building crystal. msvc is the closest you can get to "native" Windows support. Cygwin or MinGW are more focused on making code for Unix work on Windows systems, so while work on making crystal compatible with those targets might be helpful, it's likely only a small amount of work toward targeting "native" windows support. Cygwin or MinGW may be adequate.

@konovod I personally prefer Cygwin as it has a better package management system and includes mingw packages in their listing. I'm not sure how well they work, but I've had fewer problems downloading packages from Cygwin (like llvm, libgc) than building them on MinGW.

**Comment by RX14** [comment]
@omarroth does mingw allow you to create an executable file which doesn't depend on mingw? I assume cygwin would require you to depend on the cygwin DLLs in runtime. Whichever toolchain allows me to create the most native feeling binary for end-users to use is my favourite.

**Comment by omarroth** [comment]
@RX14 As far as I know, executables from both toolchains are standalone.

**Comment by sumproxy** [comment]
AFAIK cygwin has its own console and I think executables built with it will _not_ run in native windows console

**Comment by konovod** [comment]
@omarroth maybe i'm wrong (didn't used cygwin for long time), but i think mingw is actually a gcc ported to windows. It produces standalone executables that calls winapi directly. On the other hand, Cygwin executables require cygwin1.dll and they call it for any file access or sockets or threads creation. So it is a cygwin target to make porting unix programs easier (by adding a compatibility wrapper), while mingw is just working gcc, that is much better compiler (imho) than a microsoft msvc.

**Comment by sumproxy** [comment]
The Cygwin DLL is mentioned on the frontpage of Cygwin project https://www.cygwin.com/: 
> The most recent version of the Cygwin DLL is 2.5.2.

**Comment by omarroth** [comment]
@sumproxy From the [FAQ](https://www.cygwin.com/faq.html#faq.programming.static-linking): 

> Can I build a Cygwin program that does not require cygwin1.dll at runtime?

> No. If your program uses the Cygwin API, then your executable cannot run without cygwin1.dll. In particular, it is not possible to statically link with a Cygwin library to obtain an independent, self-contained executable.

So you are correct. The choice then comes down between MinGW and MSVC, whichever is more convenient for Windows devs and end-users.

**Comment by RX14** [comment]
I think we should start with MinGW, as it is easier to port to, then work on MSVC once that is complete.

**Comment by kirbyfan64** [comment]
> I think we should start with MinGW, as it is easier to port to

AFAIK the only difference for Crystal's use case would be libraries.

**Comment by RX14** [comment]
@kirbyfan64 that's a rather large difference considering the amount of libraries crystal's runtime depends on.

**Comment by dimavs** [comment]
@omarroth Just tried sieve.cr - works fine.
And about package management, msys2/mingw got perfect package management (pacman from arch linux), it's clean and simple (https://github.com/Alexpux/MSYS2-packages, https://github.com/Alexpux/MINGW-packages).

My personal preferences - msvc and mingw, cygwin is simpler (it's already got fork and fcntl implemented :-) ), but cygwin.dll as a dependence doesn't look right to me

**Comment by ysbaddaden** [comment]
MinGW is an alternative to MSVC to produce windows binaries without a POSIX layer, unlike Cygwin that bundles one (cygwin.dll). So either is fine. Maybe both MinGW and MSVC are good, because one may not know anything about MSVC and using MinGW looks simpler to UNIX users?

The Cygwin and UWin roads are a dead end eventually, because it requires another dependency, have licencing issues, yet it may help to port the compiler to Windows while we port the core/stdlib to the Windows API?

**Comment by david50407** [comment]
I think we can both support MinGW and MSVC, but start from MinGW may be easier.

**Comment by rishavs** [comment]
Hi

Any updates?

**Comment by asterite** [comment]
@rishavs not yet

**Comment by rdp** [comment]
In terms of usefulness, yes, some of us have dev boxes that run windows and it'd be nice to be able to code our crystal apps "in native windows" and even deploy them there, at times (even command line apps). Cheers!

**Comment by azah** [comment]
I got it working with the windows beta linux subsystem & bash. Just install a few things:

`sudo apt-get build-essential lib32z1-dev libssl-dev`

**Comment by exts** [comment]
@azah yeah, but that's still technically producing linux binaries, not really the same thing.

**Comment by azah** [comment]
Right, but at least it lets windows users use crystal until something native comes along.

**Comment by rishavs** [comment]
@azah 
a lot of us are actually using crystal that way. But unless we get windows compiling, I cant do anything but play around in Crystal. i am like a kid who was handed a present but told not to open it. T_T

**Comment by costajob** [comment]
Virtual machine + Vagrant?

**Comment by rishavs** [comment]
Dont know about Vagrant. will check it out.
Does it allow me to compile the binaries for a windows target?

**Comment by RX14** [comment]
@rishavs no, vagrant is just a virtual machine manager.

**Comment by rishavs** [comment]
Anything I can do to help?
I am at a beginner level, but have some free time.

If someone can list out the steps that need to be explored, i'd like to take a crack at it and report back.

**Comment by exts** [comment]
now that crystal has llvm support in 0.20.0 how hard would it be to get windows support now?

**Comment by RX14** [comment]
Crystal has always run on llvm, its the standard library which needs porting. The difficulty is that the standard library is built assuming the os is unix-based, which windows is not. The current best course of action seems to be to build using an existing "posix on windows" took chain like mingw.

**Comment by rishavs** [comment]
i really think msvc should be the way to go.
All development work in Windows is generally done using the ms toolsets.

**Comment by ysbaddaden** [comment]
@rishavs see [my comment](https://github.com/crystal-lang/crystal/issues/26#issuecomment-240356767) above: targetting Cygwin is the easiest solution to get started. It provides a POSIX command line environment (which Crystal expectes), along with a POSIX layer (which Crystal currently requires). An alternative could be to install MinGW (POSIX environment) and UWIN (POSIX layer) instead.

Once required libraries are installed or built for windows, we should be capable to cross-compile simple programs, and with some work (e.g.,: fix fiber context switch, unwinding maybe) we should be capable to port the compiler, so it runs on Windows. Then, and only then, shall we start porting the core and stdlib to the Windows API, in order to eventually get rid of the POSIX layer (Cygwin or UWIN); adding support for MSVC but not dropping support for MinGW (at least) so UNIX users can feel at home and not bother with how the hell does MSVC projects work.

I emphasize: porting to Cygwin should take a few days of work. Porting to the Windows API will require weeks of tedious hard work.

**Comment by rishavs** [comment]
@ysbaddaden 

I fear i dont understand language backends well enough to add any insight. 

Assuming Cygwin it is, how can we help? We really want Crystal to come over to Windows. Every single release, that's generally the first question out in the comments.

If you or any experienced crystal dev can break the plan into manageable tasks, we can pick them up based on our abilities.

From my side, I'll try to cross compile small programs in cygwin and report back in case of any failure.
I tried the same with llvm, but llvm turned out to be too heavy and difficult for my knowledge level.

**Comment by RX14** [comment]
The fiber context switch should be unchanged between unix and windows. It's x86_64 all the same right?

**Comment by lbguilherme** [comment]
@RX14 Actually no. The current context switch crashes on Windows because the ABI for which registers should or should not be saved is different.

---

I have some initial work on compiling to windows (Using crystal to produce `.ll`, wine clang.exe to produce `.o` and then MXE to do the linking and produce the final executable `.exe`, which can be run in wine, all done in a Linux environment. And I started to translate the stdlib to use WinAPI calls everywhere. It is very very initial, but I'll publish it in a fork this weekend so others can see/suggest/help.

**Comment by TechMagister** [comment]
Why not "just" cross-compiling using mingw64 (or mingw32) ?
-> cross compile llvm ( or download windows binaries )
-> cross compile bc ( clone repo and run mingw64-configure && mingw64-make )
-> create src/lib_c/x86_64-windows-gnu and translate stdlib
-> run mingw64-make to compile crystal

More or less :)

**Comment by RX14** [comment]
@TechMagister because without a posix emulation layer, you cannot "just" translate the stdlib. Using an existing posix emulation layer will be much easier than porting the whole stdlib to the windows API.

**Comment by shevegen** [comment]
Today I could start gvim under the (ubuntu) linux subsystem on windows.

I tried KDE Konsole and gnome-terminal but there is some dbus problem which I am unsure how to solve - but since ubuntu itself works, crystal will work too. I even could compile ruby from source in this ubuntu linux subsystem which I found funny. :D

So for the time being, the crystal project could actually say that crystal works on windows (win 10) via the linux subsystem. But I think it is also possible to compile crystal from source as well - I'll give it a try this weekend or so and try to not forget to report here.

**Comment by TsundereBug** [comment]
So, if I were to make a Crystal plugin for, say, IntelliJ, how would I handle users on Windows that don't have the Ubuntu subsystem installed? (For installing Crystal automatically.)

**Comment by RX14** [comment]
@TsundereBug the simple solution is don't support windows until we have a windows compiler out and working.

**Comment by shayneoAtNorwood** [comment]
To be blunt the "use linux-on-windows" solution is no solution at all, anymore than older "use wine on linux" hacks in other places was ever a solution. From a developer ergonomics perspective its only a step removed from "Run it in virtualbox"

**Comment by jtarchie** [comment]
There is no argument that says having native windows support would be awesome. It is just a time resource, as an OS project, delegating and prioritizing are important skillset. It helps with their own sanity, but also deliver the features they need.

The proposals for "linux-on-windows" are a great interim solution. It gives you something work with and get started with while others can work on the support for native windows. It is an open issue because this is difficult problem. Working soley on this would take the authors away from cool features, which they are providing to a substantial user base (like me), already.

For constructive feedback on this issue, saying it would be "nice" or "that is not a viable option" is not helpful. It does not help you help them accomplish your goal.

All the issuers above, spend some time and give back. Turn your "it would be nice" to "I tried and had an error". 👍

**Comment by watzon** [comment]
@shayneoAtNorwood I have to disagree as I've been running Crystal in WSL just fine and have found some workarounds, like my [wsl-proxy](https://github.com/watzon/wsl-proxy) that make it work with editors like VSCode. Not saying it's a perfect solution, but it does work in the interim.

**Comment by Wulfklaue** [comment]
Thanks @watzon  for the link to wsl-proxy. I have been wondering for a while if it is possible to communicate with WSL from outside and the wsl-proxy trick is a nice solution for this issue.

**Comment by shayneoAtNorwood** [comment]
Right, but its not really an appropriate choice for shippable code, as WSL is specifically marked as a developer only option and requires lowering certain machine security standards regarding code signing and the like to install.

**Comment by watzon** [comment]
Well yeah, but for "dipping your toes in the water" so to speak WSL works
for now. As it is Crystal is still in beta anyways, so "shippable" code is
relative.

I do agree having the compiler work on windows would be great though. I
know that there is a PR that's a WIP right now, but it's only being worked
on by one person and could probably use some other contributors.

**Comment by Wulfklaue** [comment]
> Right, but its not really an appropriate choice for shippable code, as WSL is specifically marked as a developer only option and requires lowering certain machine security standards regarding code signing and the like to install. For coder stuff, its fine. 

That might have been true a few weeks ago, but with Windows Fall Update, WSL is now a full featured option like hyper-v... No more beta, no more need to drop to "developer" mode.

https://blogs.msdn.microsoft.com/commandline/2017/10/11/whats-new-in-wsl-in-windows-10-fall-creators-update/

> But I could not in good conscience ship something that relies on the windows linux subsystem because it would require them to put their machine into "developer" mode and reduce the effectiveness of windows built in hardening.

Shipping a final product on a language that is 0.23, is not exactly the smartest thing to do. Crystal is usable for your own development and your own hosting ( where you are in control of the environment ) but shipping a product written in a beta language, that is asking for trouble.

**Comment by ysbaddaden** [comment]
**Please stop flooding about WSL!**

This issue is about **adding support for Windows* not about running crystal in whatever VM/emulation system.

We gradually had support, but Windows is still UNSUPPORTED.

**Comment by exts** [comment]
Thank you @ysbaddaden, no windows users care about VM's we want native support.

**Comment by SolarLune** [comment]
Yo, just wanted to pop in and pipe up. As a game developer, despite the fact that I like to work on Linux and Crystal's looking pretty great, most gamers are on Windows. So, Windows support would be needed for a supported target platform for me to be able to use Crystal for development.

**Comment by RX14** [comment]
Please, can we keep the comments to *constructive* comments with discussion about actually implementing this feature. We know how much the community wants this, we don't need any more comments which essentially translate to 👍.

**Comment by SolarLune** [comment]
I do agree with you, but considering that it's not an insignificant portion of people that believe that running Crystal on a bash VM is OK for now, it can't hurt to bring up another use case in which a fully compatible version of Crystal is absolutely necessary.

**Comment by mverzilli** [comment]
I second @ysbaddaden and @RX14. Please, stop asking for something that was already accepted and is one of the top project goals. If it hasn't happened just yet, it is because of lack of resources and time. It is slowly progressing though. 

Verbalizing more requests for it to happen will only make this thread harder to follow and confusing for by-passers. Learn to take "yes" for an answer! :)

**Comment by sam0x17** [comment]
One more thing to consider: the second we get native windows support, there are a lot of people who are going to be interested in doing something like Electron but with crystal -- I have several friends working on desktop apps in Electron who would switch to crystal in a second if the cross platform support was there. 

Not only that, but crystal is _perfect_ for doing little one-off command line utility programs, and windows support would just make that sweeter. Right now the best I can find as a cross platform solution for that is Node.js with `node-pkg` (or electron).

**Comment by RX14** [comment]
#5339 has just been merged, which brings the ability to compile `puts "Hello World"` on native windows. Over the next few months it's likely people (including me) will PR in new support for more and more of the crystal standard library including networking, fibers, and evented IO. If people want to help with that effort, or want to "take" a piece of that work to work on themself, please comment below.

**Comment by TsundereBug** [comment]
I'm not against implementing concurrency on Windows, however I may not have the time or knowledge of current Crystal to do so.

When is this expected to be fully complete via development like this?

**Comment by RX14** [comment]
@TsundereBug that's hard to say, we'll never get "fully complete" since we can't implement `fork()`. It could be as low as a few months or as long as a year to get to "production ready" though.

**Comment by sam0x17** [comment]
@RX14 that is actually super exciting 🕺

**Comment by RX14** [comment]
Just FYI, windows coordination is now happening in #5430. Not sure when this issue should be closed, I guess most people would want to at least see a self-hosted compiler before closing this issue.

**Comment by Snake-Sanders** [comment]
I am quite interesting in the potential of this language, I know a bit of many languages but I couldn't say how I can help you here. My proposal is to create lots of small issues or tasks so people like me can take a piece and contribute. This will allow more people to contribute as well. The more we are, the better.

**Comment by svnpenn** [comment]
https://github.com/crystal-lang/crystal/issues/6170

**Comment by HertzDevil** [comment]
The Windows package is now self-hosting and rather portable. We are targetting MSVC and the Win32 APIs directly, instead of Cygwin or MinGW(-x64) (#6170).

I don't think there are any immediate plans to be able to build DLLs on Windows, nor build libraries on Unix-like systems, for that matter. DLL support as a whole has many details yet to be decided (#11575).

**Comment by Neltherion** [comment]
It's been nearly a year and a half later... Any news?

**Comment by tjpalmer** [comment]
See #5430 for latest updates, as I understand it.

**Comment by straight-shoota** [comment]
By now all the main platform features for Windows (#5430) are finished! 🚀 

But there are still some smaller stories pending to complete Windows support.
We have created a project board to keep track of them: https://github.com/orgs/crystal-lang/projects/11/views/5
Contributions are welcome!

If you want to financially support the ongoing efforts, check out the [Windows suppport](https://opencollective.com/crystal-lang/projects/windows-support) project on Open Collective. It's a dedicated contribution to boost the development of Windows support, not a subscription. Thank you very much 🙇

**Comment by HertzDevil** [comment]
The Windows project board and the other open issues are sufficient to track the progress of Windows support. #921 is a more appropriate place to discuss building DLLs as mentioned in OP.

Closing 🎊
